### ISSUE TITLE
False positive "Comparison of String objects using == or != in ..." warning.

### ISSUE THREAD

**Issue by slavap** [spotbugs/spotbugs#189]
```java
String oldValue = getValue();
// some calculations, which may change value
String newValue = getValue();
boolean eq = newValue == oldValue || newValue != null && newValue.equals(oldValue);
```
This pattern is absolutely clear, no need to raise warning in such cases.

**Comment by KengoTODA** [comment]
I personally think that SpotBugs does not have to support this request.

It is technically possible to handle this case (reference comparison which invokes `#equals(Object)` with short-circuit evaluation) as "it would be intentional operation", however this idiom is not good practice nowadays, and nobody can ensure that it is always intentional. So IMHO supporting this request just introduces complexity and makes analysis slow.

@slavap I recommend you to replace your logic with [`Objects.equals(Object, Object)`](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#equals-java.lang.Object-java.lang.Object-) introduced from Java7, then you can make your code short and clear. It is also easy to improve test coverage. If you're working with Java 1.6 or older, `@SuppressFBWarning` is the last option.

**Comment by slavap** [comment]
@KengoTODA I know how to replace this code with Objects.equals(), but ironically it's implemented the same way as my initial complaint :-) IMO good static analyzer should not give false alarms on simple and absolutely valid code (it looks strange and reduce overall perception of this tool), and I just don't see how it could slow down analysis.

**Comment by KengoTODA** [comment]
Yes, it's totally same with your implementation. So you can use it to hide that unintuitive complexity behind of library.

**Comment by iloveeclipse** [comment]
What FindBugs means here is that is highly uncommon to compare *Strings* using == operator. In many cases this is just a bug, because equal Strings could belong to different pointers. In this particular case FindBugs is not clever enough to see that this is not a bug, but because there is a simple workaround for the code, I would close it as won't fix.
